chunk/aws_storage_client.go

```diff
@@ -61,41 +67,86 @@ var (
 		Name:      "dynamo_failures_total",
 		Help:      "The total number of errors while storing chunks to the chunk store.",
 	}, []string{tableNameLabel, errorReasonLabel})
+	s3RequestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "s3_request_duration_seconds",
+		Help:      "Time spent doing S3 requests.",
+		Buckets:   []float64{.025, .05, .1, .25, .5, 1, 2},
+	}, []string{"operation", "status_code"})
 )
 
 func init() {
 	prometheus.MustRegister(dynamoRequestDuration)
 	prometheus.MustRegister(dynamoConsumedCapacity)
 	prometheus.MustRegister(dynamoFailures)
+	prometheus.MustRegister(s3RequestDuration)
 }
 
-type dynamoClientAdapter struct {
-	DynamoDB dynamodbiface.DynamoDBAPI
+// DynamoDBConfig specifies config for a DynamoDB database.
+type DynamoDBConfig struct {
+	DynamoDB util.URLValue
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *DynamoDBConfig) RegisterFlags(f *flag.FlagSet) {
+	f.Var(&cfg.DynamoDB, "dynamodb.url", "DynamoDB endpoint URL with escaped Key and Secret encoded. "+
+		"If only region is specified as a host, proper endpoint will be deduced. Use inmemory:///<table-name> to use a mock in-memory implementation.")
+}
+
+// AWSStorageConfig specifies config for storing data on AWS.
+type AWSStorageConfig struct {
+	DynamoDBConfig
+	S3 util.URLValue
 }
 
-// NewDynamoDBClient makes a new DynamoDBClient
-func NewDynamoDBClient(dynamoDBURL string) (StorageClient, string, error) {
-	url, err := url.Parse(dynamoDBURL)
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *AWSStorageConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.DynamoDBConfig.RegisterFlags(f)
+	f.Var(&cfg.S3, "s3.url", "S3 endpoint URL with escaped Key and Secret encoded. "+
+		"If only region is specified as a host, proper endpoint will be deduced. Use inmemory:///<bucket-name> to use a mock in-memory implementation.")
+}
+
+type awsStorageClient struct {
+	DynamoDB   dynamodbiface.DynamoDBAPI
+	S3         s3iface.S3API
+	bucketName *string
```

Comment:
ah, just to not do this weird `aws.String(..)` thing.. I would vote for making it just string, no need to keeping ptr on our side. It just surprises. Not a blocker.